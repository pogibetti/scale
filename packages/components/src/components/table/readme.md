# scale-table



<!-- Auto Generated Below -->


## Properties

| Property      | Attribute      | Description                    | Type                                     | Default     |
| ------------- | -------------- | ------------------------------ | ---------------------------------------- | ----------- |
| `customClass` | `custom-class` | (optional) Tag class           | `string`                                 | `''`        |
| `styles`      | --             | (optional) Injected jss styles | `StyleSheet<string \| number \| symbol>` | `undefined` |
| `variant`     | `variant`      | (optional) Visual variant      | `"compressed" \| "regular"`              | `'regular'` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
