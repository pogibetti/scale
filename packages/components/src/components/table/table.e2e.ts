import { newE2EPage } from '@stencil/core/testing';

describe('scale-table', () => {
  let page;
  beforeEach(async () => {
    page = await newE2EPage();
    await page.setContent(`
      <html>
        <head></head>
        <body>
          <script>
            function sort(element) {
              const prevSort = element.getAttribute('aria-sort');
              const newSort = prevSort === 'descending' ? 'ascending' : prevSort === 'ascending' ? 'none' : 'descending';
              element.setAttribute('aria-sort', newSort)
            }
          </script>

          <scale-table>
          <table>
            <thead>
              <tr>
                <th id="title-header" onclick="sort(this)">Title</th>
                <th>Time</th>
                <th>Euros</th>
              </tr>
            </thead>
            <tbody>
            <tr>
              <td>University of Plymouth</td>
              <td>00:00:20</td>
              <td>100.245,10</td>
            </tr>
            <tr>
              <td>University of Plymouth</td>
              <td>00:00:20</td>
              <td>100.245,10</td>
            </tr>
            <tr>
              <td>University of Plymouth</td>
              <td>00:00:20</td>
              <td>100.245,10</td>
            </tr>
            </tbody>
            <tfoot>
              <tr>
                <td>Total</td>
                <td />
                <td />
                <td>00:00:20</td>
                <td>100.245,10</td>
              </tr>
            </tfoot>
          </table>
        </scale-table>
        </body>
      </html>
    `);
  });

  it('should not render sort icon initially', async () => {
    expect(await page.find('svg')).toMatchInlineSnapshot(`null`);
  });

  it('should render descending sort icon once table head is clicked', async () => {
    await page.click('#title-header');
    await page.waitForChanges();

    expect(
      (await page.find('polygon')).getAttribute('fill')
    ).toMatchInlineSnapshot(`"#CDCDCD"`);
  });

  it('should render ascending sort icon when table head is clicked twice', async () => {
    await page.click('#title-header');
    await page.click('#title-header');
    await page.waitForChanges();

    expect(
      (await page.find('polygon')).getAttribute('fill')
    ).toMatchInlineSnapshot(`"#000000"`);
  });

  it('should not render a sort icon when clicked the third time', async () => {
    await page.click('#title-header');
    await page.click('#title-header');
    await page.click('#title-header');
    await page.waitForChanges();

    expect(await page.find('svg')).toMatchInlineSnapshot(`null`);
  });
});
