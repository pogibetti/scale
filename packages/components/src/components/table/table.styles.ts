import { JssStyle } from 'jss';

export const styles: JssStyle = {
  table: {
    '& table': {
      fontFamily: ({ typography }) => typography.fontFamily,
      borderSpacing: 0,
    },
    '& th': {
      background: ({ colors }) => colors.grey[100],
      color: ({ colors }) => colors.grey[600],
      lineHeight: '32px',
      fontSize: 12,
      textAlign: 'left',
      cursor: 'pointer',

      '&[aria-disabled="true"]': {
        cursor: 'default',
        pointerEvents: 'none',
      },
    },
    '& th, td': {
      padding: '0 8px',
    },
    '& td': {
      padding: '16px 8px',
      borderBottom: '1px solid #DDDDDD',
    },

    '& tfoot tr td': {
      fontWeight: 600,
      padding: '8px',
      borderBottom: '1px solid black',
    },
  },

  'table--variant-compressed': {
    '& td': {
      padding: '8px',
    },
  },
};
