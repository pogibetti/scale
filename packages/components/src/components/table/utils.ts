export type TDirection =
  | 'ascending'
  | 'descending'
  | 'none'
  | 'other'
  | ''
  | undefined;

export const SORT_INDICATOR_ID = 'scale-sort-indicator';

export const getSortIndicator = (direction: TDirection) =>
  `<span id="${SORT_INDICATOR_ID}">
        ${
          !direction || ['none', 'other'].includes(direction)
            ? '<span />'
            : `<svg  width="6px" height="9px" viewBox="0 0 6 9" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
                <g id="${SORT_INDICATOR_ID}-1" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
                  <g id="${SORT_INDICATOR_ID}-2" transform="translate(-183.000000, -7310.000000)">
                    <g id="${SORT_INDICATOR_ID}-3" transform="translate(174.000000, 6631.000000)">
                      <g id="${SORT_INDICATOR_ID}-4" transform="translate(0.000000, 668.000000)">
                        <g id="${SORT_INDICATOR_ID}-5">
                          <polygon id="Rectangle" fill="${
                            direction === 'ascending' ? '#000000' : '#CDCDCD'
                          }" transform="translate(12.000000, 14.000000) rotate(45.000000) translate(-12.000000, -14.000000) " points="10 12 14 12 10 16"></polygon>
                          <polygon id="Rectangle" fill="${
                            direction === 'ascending' ? '#CDCDCD' : '#000000'
                          }" transform="translate(12.000000, 17.000000) scale(1, -1) rotate(45.000000) translate(-12.000000, -17.000000) " points="10 15 14 15 10 19"></polygon>
                        </g>
                      </g>
                    </g>
                  </g>
                </g>
              </svg>`
        }
    </span>
`;
