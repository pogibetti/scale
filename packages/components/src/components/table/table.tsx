import { Component, Prop, h, Host, Element, State } from '@stencil/core';
import { CssClassMap } from '../../utils/utils';
import classNames from 'classnames';
import { styles } from './table.styles';
import { CssInJs } from '../../utils/css-in-js';
import { StyleSheet } from 'jss';
import Base from '../../utils/base-interface';
import { getSortIndicator, TDirection, SORT_INDICATOR_ID } from './utils';

@Component({
  tag: 'scale-table',
  shadow: true,
})
export class Table implements Base {
  @Element() hostElement: HTMLElement;
  /** (optional) Tag class */
  @Prop() customClass?: string = '';
  /** (optional) Visual variant */
  @Prop() variant?: 'regular' | 'compressed' = 'regular';
  /** (optional) Injected jss styles */
  @Prop() styles?: StyleSheet;
  /** decorator Jss stylesheet */
  @CssInJs('Table', styles) stylesheet: StyleSheet;
  @State() observer;

  disconnectedCallback() {
    if (this.observer) {
      this.observer.disconnect();
    }
  }

  componentWillLoad() {
    const tableChildren = this.hostElement.children[0];

    tableChildren.querySelectorAll('th').forEach(th => {
      if (typeof MutationObserver !== 'undefined') {
        this.observer = new MutationObserver(function(mutations) {
          mutations.forEach(function(mutation) {
            const mutationTarget = mutation.target as HTMLElement;
            if (
              mutation.type === 'attributes' &&
              mutation.attributeName === 'aria-sort'
            ) {
              const sortIndicators = tableChildren.querySelectorAll(
                `#${SORT_INDICATOR_ID}`
              );

              sortIndicators.forEach(sortIndicator => {
                sortIndicator.parentNode.removeChild(sortIndicator);
              });

              mutationTarget.insertAdjacentHTML(
                'afterbegin',
                getSortIndicator(
                  mutationTarget.getAttribute('aria-sort') as TDirection
                )
              );
            }
          });
        });

        this.observer.observe(th, {
          attributes: true,
        });
      }
    });
  }

  componentWillUpdate() {}

  render() {
    const tableChildren = this.hostElement.children[0];

    tableChildren.querySelectorAll('th').forEach(th => {
      const sort = th.getAttribute('aria-sort') as TDirection;
      th.insertAdjacentHTML('afterbegin', getSortIndicator(sort));
    });

    return (
      <Host class={this.getCssClassMap()}>
        <style>{this.stylesheet.toString()}</style>
        <slot />
      </Host>
    );
  }

  getCssClassMap(): CssClassMap {
    const { classes } = this.stylesheet;
    return classNames(
      classes.table,
      this.customClass && this.customClass,
      this.variant && classes[`table--variant-${this.variant}`]
    );
  }
}
