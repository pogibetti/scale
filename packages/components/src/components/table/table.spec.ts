import { newSpecPage } from '@stencil/core/testing';
import { Table } from './table';
import { styles } from './table.styles';
import jss from 'jss';

describe('Table', () => {
  let element;
  let stylesheet;

  beforeEach(async () => {
    element = new Table();
    stylesheet = element.stylesheet = jss.createStyleSheet(styles as any);
  });

  it('should match snapshot', async () => {
    // @ts-ignore
    global.MutationObserver = class MutationObserver {
      observe() {}
      disconnect() {}
    };
    const page = await newSpecPage({
      components: [Table],
      html: `
        <scale-table>
          <table>
            <thead>
              <tr>
                <th aria-sort="descending">Title</th>
                <th>Time</th>
                <th>Euros</th>
              </tr>
            </thead>
            <tbody>
            <tr>
              <td>University of Plymouth</td>
              <td>00:00:20</td>
              <td>100.245,10</td>
            </tr>
            <tr>
              <td>University of Plymouth</td>
              <td>00:00:20</td>
              <td>100.245,10</td>
            </tr>
            <tr>
              <td>University of Plymouth</td>
              <td>00:00:20</td>
              <td>100.245,10</td>
            </tr>
            </tbody>
            <tfoot>
              <tr>
                <td>Total</td>
                <td />
                <td />
                <td>00:00:20</td>
                <td>100.245,10</td>
              </tr>
            </tfoot>
          </table>
        </scale-table>
      `,
    });
    expect(page.root).toMatchSnapshot();

    // @ts-ignore
    global.MutationObserve = undefined;
  });

  it('should handle css classes', () => {
    element.customClass = 'custom';
    expect(element.getCssClassMap()).toContain('custom');

    element.variant = 'compressed';
    stylesheet.addRule('table--variant-compressed', {});
    expect(element.getCssClassMap()).toContain(
      stylesheet.classes['table--variant-compressed']
    );
  });
});
