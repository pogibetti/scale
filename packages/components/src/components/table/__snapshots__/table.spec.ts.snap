// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should match snapshot 1`] = `
<scale-table class="table-0-0-3">
  <mock:shadow-root>
    <style>
      .table-0-0-3 {}
.table-0-0-3 table {
  border-spacing: 0;
  font-family: 'Roboto', 'Helvetica', 'Arial', sans-serif;
}
.table-0-0-3 th {
  cursor: pointer;
  font-size: 12px;
  text-align: left;
  line-height: 32px;
  background: #F5F5F5;
  color: #757575;
}
.table-0-0-3 th, .table-0-0-3 td {
  padding: 0 8px;
}
.table-0-0-3 td {
  padding: 16px 8px;
  border-bottom: 1px solid #DDDDDD;
}
.table-0-0-3 tfoot tr td {
  padding: 8px;
  font-weight: 600;
  border-bottom: 1px solid black;
}
.table-0-0-3 th[aria-disabled="true"] {
  cursor: default;
  pointer-events: none;
}
.table--variant-compressed-0-0-4 {}
.table--variant-compressed-0-0-4 td {
  padding: 8px;
}
    </style>
    <slot></slot>
  </mock:shadow-root>
  <table>
    <thead>
      <tr>
        <th aria-sort="descending">
          <span id="scale-sort-indicator">
            <svg height="9px" version="1.1" viewBox="0 0 6 9" width="6px" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns="http://www.w3.org/2000/svg">
              <g fill="none" fill-rule="evenodd" id="scale-sort-indicator-1" stroke="none" stroke-width="1">
                <g id="scale-sort-indicator-2" transform="translate(-183.000000, -7310.000000)">
                  <g id="scale-sort-indicator-3" transform="translate(174.000000, 6631.000000)">
                    <g id="scale-sort-indicator-4" transform="translate(0.000000, 668.000000)">
                      <g id="scale-sort-indicator-5">
                        <polygon fill="#CDCDCD" id="Rectangle" points="10 12 14 12 10 16" transform="translate(12.000000, 14.000000) rotate(45.000000) translate(-12.000000, -14.000000) "></polygon>
                        <polygon fill="#000000" id="Rectangle" points="10 15 14 15 10 19" transform="translate(12.000000, 17.000000) scale(1, -1) rotate(45.000000) translate(-12.000000, -17.000000) "></polygon>
                      </g>
                    </g>
                  </g>
                </g>
              </g>
            </svg>
          </span>
          Title
        </th>
        <th>
          <span id="scale-sort-indicator">
            <span></span>
          </span>
          Time
        </th>
        <th>
          <span id="scale-sort-indicator">
            <span></span>
          </span>
          Euros
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          University of Plymouth
        </td>
        <td>
          00:00:20
        </td>
        <td>
          100.245,10
        </td>
      </tr>
      <tr>
        <td>
          University of Plymouth
        </td>
        <td>
          00:00:20
        </td>
        <td>
          100.245,10
        </td>
      </tr>
      <tr>
        <td>
          University of Plymouth
        </td>
        <td>
          00:00:20
        </td>
        <td>
          100.245,10
        </td>
      </tr>
    </tbody>
    <tfoot>
      <tr>
        <td>
          Total
        </td>
        <td></td>
        <td></td>
        <td>
          00:00:20
        </td>
        <td>
          100.245,10
        </td>
      </tr>
    </tfoot>
  </table>
</scale-table>
`;
